components/tikv_util/src/metrics/threads_linux.rs

```diff
@@ -287,6 +289,139 @@ lazy_static! {
     };
 }
 
+#[inline]
+fn get_name(command: &str) -> String {
+    if command != "" {
+        return command.to_owned();
+    }
+    String::from("anony")
+}
+
+fn collect_metrics_by_name(
+    names: &HashMap<i32, String>,
+    values: &HashMap<i32, f64>,
+) -> HashMap<String, u64> {
+    let mut new_map: HashMap<String, u64> = HashMap::default();
+    for (tid, name) in names {
+        let new_value = new_map.entry(name.to_string()).or_insert(0);
+        if let Some(value) = values.get(&tid) {
+            *new_value += *value as u64;
+        }
+    }
+    new_map
+}
+
+#[inline]
+fn update_metric(metrics: &mut HashMap<i32, f64>, tid: i32, new_value: f64) -> f64 {
+    let value = metrics.entry(tid).or_insert(0.0);
+    let mut delta = new_value - *value;
+    if delta > 0.0 {
+        *value = new_value;
+    } else {
+        delta = 0.0
+    }
+    delta
+}
+
+#[inline]
+fn update_rate(rates: &mut HashMap<i32, f64>, tid: i32, value_delta: f64, time_delta: f64) {
+    let value = rates.entry(tid).or_insert(0.0);
+    if value_delta > 0.0 && time_delta > 0.0 {
+        *value = value_delta / time_delta
+    }
+}
+
+#[derive(Default)]
+struct ThreadMetrics {
+    cpu_times: HashMap<i32, f64>,
+    read_ios: HashMap<i32, f64>,
+    write_ios: HashMap<i32, f64>,
+}
+
+impl ThreadMetrics {
+    fn clear(&mut self) {
+        self.cpu_times.clear();
+        self.read_ios.clear();
+        self.write_ios.clear();
+    }
+}
+
+/// Use to collect cpu usages and disk I/O rates
+pub struct ThreadInfoStatistics {
+    pid: pid_t,
+    last_instant: Instant,
+    tid_names: HashMap<i32, String>,
+    metrics_rate: ThreadMetrics,
+    metrics_total: ThreadMetrics,
+}
+
+impl ThreadInfoStatistics {
+    pub fn new() -> Self {
+        let pid = unsafe { libc::getpid() };
+
+        let mut thread_stats = ThreadInfoStatistics {
+            pid,
+            last_instant: Instant::now(),
+            tid_names: HashMap::default(),
+            metrics_rate: ThreadMetrics::default(),
+            metrics_total: ThreadMetrics::default(),
+        };
+        thread_stats.record();
+        thread_stats
+    }
+
+    pub fn record(&mut self) {
+        let current_instant = Instant::now();
+        let time_delta = (current_instant - self.last_instant).as_millis() as f64 / 1000.0;
+        self.last_instant = current_instant;
+
+        self.metrics_rate.clear();
+
+        let tids = get_thread_ids(self.pid).unwrap();
+        for tid in tids {
+            if let Ok(stat) = pid::stat_task(self.pid, tid) {
+                let name = get_name(&stat.command);
+                self.tid_names.entry(tid).or_insert(name);
+
+                let cpu_time = cpu_total(&stat) * 100.0;
+                let cpu_delta = update_metric(&mut self.metrics_total.cpu_times, tid, cpu_time);
```

Comment:
Can we merge these two update functions?

Reply:
OK